test_runner/src/main/kotlin/ftl/json/SavedMatrix.kt

```diff
@@ -3,118 +3,122 @@ package ftl.json
+    val clientDetails: Map<String, String>?,
+    val gcsPathWithoutRootBucket: String,
+    val gcsRootBucket: String,
+    val webLinkWithoutExecutionDetails: String?
+)
+
+fun createSavedMatrix(testMatrix: TestMatrix) = defaultSavedMatrix().updateWithMatrix(testMatrix)
 
-    var billableVirtualMinutes: Long = 0
-        private set
-    var billablePhysicalMinutes: Long = 0
-        private set
-    var outcome: String = ""
-        private set
-    var outcomeDetails: String = ""
-        private set
-    val clientDetails = matrix.clientInfo?.clientInfoDetails
-        ?.map { it.key to it.value }
-        ?.toMap()
-
-    init {
-        if (this.state == FINISHED) finished(matrix)
-    }
-
-    // ExitCodeFromRollupOutcome - https://github.com/bootstraponline/gcloud_cli/blob/137d864acd5928baf25434cf59b0225c4d1f9319/google-cloud-sdk/lib/googlecloudsdk/api_lib/firebase/test/exit_code.py#L46
-    fun failed(): Boolean {
-        // outcome of the test execution.
-        // skipped means unsupported environment
-        return when (outcome) {
-            failure -> true
-            skipped -> true
-            inconclusive -> true
-            else -> false
-        }
-    }
+fun SavedMatrix.canceledByUser() = outcomeDetails == ABORTED_BY_USER_MESSAGE
 
-    /** return true if the content changed **/
-    fun update(matrix: TestMatrix): Boolean {
-        val newState = matrix.state
-        val newLink = matrix.webLink()
-        val changedState = state != newState
-        val changedLink = webLink != newLink
+fun SavedMatrix.infrastructureFail() = outcomeDetails == INFRASTRUCTURE_FAILURE_MESSAGE
 
-        if (changedState) {
-            updateState(newState, matrix)
-        }
+fun SavedMatrix.incompatibleFail() = outcomeDetails in arrayOf(
+    INCOMPATIBLE_APP_VERSION_MESSAGE,
+    INCOMPATIBLE_ARCHITECTURE_MESSAGE,
+    INCOMPATIBLE_DEVICE_MESSAGE
+)
+
+fun SavedMatrix.isFailed() = when (outcome) {
+    failure -> true
+    skipped -> true
+    inconclusive -> true
+    else -> false
+}
 
-        if (changedLink) {
-            this.webLink = newLink
-        }
+fun SavedMatrix.needsUpdate(newMatrix: TestMatrix): Boolean {
+    val newState = newMatrix.state
+    val newLink = newMatrix.webLink()
+    val changedState = state != newState
+    val changedLink = webLink != newLink
+    return (changedState || changedLink)
+}
 
-        return changedState || changedLink
-    }
+internal fun SavedMatrix.updateWithMatrix(newMatrix: TestMatrix): SavedMatrix {
+    return if (needsUpdate(newMatrix)) {
+        return updatedSavedMatrix(newMatrix)
+    } else this
+}
 
-    private fun updateState(newState: String, testMatrix: TestMatrix) {
-        state = newState
-        when (state) {
-            FINISHED -> finished(testMatrix)
+private fun SavedMatrix.updatedSavedMatrix(newMatrix: TestMatrix): SavedMatrix {
+    var outcomeDetails = if (this.outcomeDetails.isNullOrEmpty()) "" else this.outcomeDetails
```

Comment:
```SavedMatrix.outcomeDetails``` is non-null property so we can change this line to

```

var outcomeDetails = this.outcomeDetails

```

Reply:
The change to immutability as shown that one test causes some issues with the outcome details. Infact i made some changes as there is one small case where it can be null. So i have changed it to be nullable fixed the tests that now fail.